Add GitCommentHelpHandler class

betulependule opened the pull request:
A new job handler has been created so that `packit-service` now has the ability to respond to the `help` command used in issue and PR comments. A comment is then added to the issue / PR, which contains a help message that includes the usage, link to documentation and contacts. Three tests are included to ensure functionality.

<!-- TODO list -->

TODO:

- [x] Update or write new documentation in `packit/packit.dev`.
- [x] In the help message, add reference to `/packit-ci` when user uses `/packit` and the other way around. This is in case the user uses `/packit` when they should be using `/packit-ci` and vice versa.

<!-- notes for reviewers -->

<!-- Links to other issues or pull requests,
     for cross-repository links use: ‹namespace›/‹repository›#‹ID of issue›
       (‹namespace›/‹repository›!‹ID of PR› respectively)
-->

Fixes [#1668](https://github.com/packit/packit-service/issues/1668)

Related to

Merge after [#2850](https://github.com/packit/packit-service/pull/2850)

<!-- release notes footer -->

RELEASE NOTES BEGIN

Packit-service now supports the `help` command in issue and PR comments.

RELEASE NOTES END

betulependule commented:
/gemini review

betulependule commented on packit_service/worker/handlers/forges.py:
```diff
@@ -279,3 +296,61 @@ def verify(self, namespace: str, fas_account: str) -> TaskResults:
+        # sys.stdout = help_message_buffer
+        # parser.print_help()
+        # sys.stdout = backup_stdout
+        # help_message = help_message_buffer.getvalue()
```
Based on the [docs](https://docs.python.org/3/library/contextlib.html#contextlib.redirect_stdout), `contextlib.redirect_stdout` isn't thread-safe.

> Note that the global side effect on [sys.stdout](https://docs.python.org/3/library/sys.html#sys.stdout) means that this context manager is not suitable for use in library code and most threaded applications. It also has no effect on the output of subprocesses. However, it is still a useful approach for many utility scripts.

A different solution is needed here.

betulependule commented on packit_service/worker/handlers/forges.py:
```diff
@@ -279,3 +295,54 @@ def verify(self, namespace: str, fas_account: str) -> TaskResults:
+        #     self.comment, # type: ignore
+        #     self.service_config.comment_command_prefix,
+        # )
+        # if self.comment.startswith("/packit-ci"): # type: ignore
```
There should be no need to check that the command used was `help` as that is already being checked in `SteveJobs` using the `is_help_comment` function. Both `/packit-ci` and `/packit-ci-stg` require the  parser returned by `get_pr_comment_parser_fedora_ci`. The change above would cause `/packit-ci-stg` to use the wrong parser.

betulependule commented on packit_service/worker/jobs.py:
```diff
@@ -1116,6 +1128,28 @@ def is_fas_verification_comment(self, comment: str) -> bool:
+            packit_comment_command_prefix,
+        )
+
+        return bool(command and command[0] == PACKIT_HELP_COMMAND)
```
The "-stg" suffixes definitely need to be taken into account. However, I don't think I like the trial and error for loop. I'll try a different solution.

betulependule commented:
/gemini review

betulependule commented:
I'm thinking whether it wouldn't be better to remove the `get_pr_comment_parser` and `get_pr_comment_parser_fedora_ci` and instead, implement two singleton classes `PRCommentParser` and `PRCommentParserFedoraCI` and move them from `utils.py` to a dedicated file.

majamassarini commented on packit_service/worker/jobs.py:
```diff
@@ -322,6 +327,15 @@ def process(self) -> list[TaskResults]:
             # on the issue created by us or not in packit/notifications?
+        elif isinstance(
+            self.event,
+            (github.pr.Comment, gitlab.mr.Comment, pagure.pr.Comment),
```
We also react to comments in issues. So, I think we should do the same also for comments in issues.

majamassarini commented on packit_service/worker/handlers/forges.py:
```diff
@@ -279,3 +297,47 @@ def verify(self, namespace: str, fas_account: str) -> TaskResults:
+        self.comment = self.data.event_dict.get("comment")
+
+    def run(self) -> TaskResults:
+        if self.comment.startswith("/packit-ci"):  # type: ignore
```
```suggestion
        if self.comment.startswith("/packit-ci") or self.comment.startswith("/packit-ci-stg") :  # type: ignore
```

lbarcziova reviewed (commented):
overall looks neat!

betulependule commented on packit_service/worker/handlers/forges.py:
```diff
@@ -279,3 +297,47 @@ def verify(self, namespace: str, fas_account: str) -> TaskResults:
+            parser = get_pr_comment_parser_fedora_ci(
+                prog=HELP_COMMENT_PROG_FEDORA_CI,
+                description=HELP_COMMENT_DESCRIPTION,
+                epilog=HELP_COMMENT_EPILOG.format(note=HELP_NOTE_FEDORA_CI),
```
Yes, it would probably be better to have "/packit help" use the keyword "/packit" and have "/packit-stg help" instead use "/packit-stg".  Simply using "/packit | /packit-stg" in both cases would be probably confusing to users. I'll improve this.

majamassarini commented on packit_service/worker/handlers/forges.py:
```diff
@@ -279,3 +300,68 @@ def verify(self, namespace: str, fas_account: str) -> TaskResults:
+@reacts_to(event=gitlab.issue.Comment)
+class GitPullRequestHelpHandler(
+    JobHandler,
+    GetPullRequestMixin,
```
I am afraid you can't use the `GetPullRequestMixin` when reacting to an issue comment. I would also create some tests with an issue comment event to see how it reacts.

betulependule commented:
I didn't want to make it hard to see the most recent changes since the last review, so I haven't squished the commits yet. I'll do so before merge once it is ready.
